keps/sig-instrumentation/1668-log-tracking/README.md

```diff
@@ -0,0 +1,372 @@
+### Summary of Cases
+
+ - Given a component log(such as error log), find the API request that caused this (error) log.
+ - Given an API Request(such as suspicious API request), find the resulting component logs.
+
+### Logging metadata
+
+We use three logging meta-data, and propagate them each K8s component by using OpenTelemetry.
+OpenTelemetry has SpanContext and [Baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md) which is used for propagation of K8s component.
+
+| meta-data name | feature |
+| ------ | ------ |
+| traceid | We use SpanContext.TraceID as traceid<br>traceid spans an user request.<br>traceid is unique for user's request |
+| spanid | We use SpanContext.SpanID as spanid<br>spanid spans a controller action.<br>spanid is unique for controller action |
+| initialtraceid | We implement new id(InitialTraceID) to SpanContext<br>We use [Baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md) to propagate initialtraceid<br/>We use UID of root object as initialtraceid<br>initialtraceid spans the entire object lifecycle. <br>initialtraceid is unique for related objects |
+
+All of three id's inception is from object creation and it dies with object deletion
+
+## Design Details
+
+### Prerequisite
+
+We need to consider three cases:
+- Case1: Requests from kubectl that creating an object
+- Case2: Requests from kubectl other than creating (e.g. updating, deleting) an object
+- Case3: Requests from controllers
+
+The design below is based on the above three cases
+
+The following picture show our design![design](./overview.png)
+
+
+
+we don't have any modifications in kubectl in this design.
+
+### API Server Tracing
+
+**1. refer to [KEP647](https://github.com/kubernetes/enhancements/issues/647)**
+
+KEP647 will help to
+
+```
+Goals
+The API Server generates and exports spans for incoming and outgoing requests.
+The API Server propagates context from incoming requests to outgoing requests.
+```
+
+**2. [Make up a new outgoing request](#design-of-id-progagation-in-client-go)**
+
+### Design of ID propagation (controller)
+
+**1. Propagate golang ctx from objects to API Calls**
+
+When controllers create/update/delete an object A based on another B, we propagate context from B to A. E.g.:
+```
+    ctx = httptrace.SpanContextFromAnnotations(context.Background(), objB.GetAnnotations())
+    err = r.KubeClient.CoreV1().Create(ctx, objA...)
+```
+We do propagation across objects without adding traces to that components.
+
+SpanContextFromAnnotations will do something like below pseudo code
+```
+func SpanContextFromAnnotations(){
+	// 1. decode annotation to SpanContext
+	// 2. generate Span from SpanContext
+	// 3. chain Span to ctx which is required by propagators.Inject
+}
+```
+**2. [Make up a new outgoing request](#design-of-id-progagation-in-client-go)**
```

Comment:
If a user is not using the admission controller, what would the effect be of passing context here?  Do we expect to do logging in the API Server?

Reply:
> If a user is not using the admission controller, what would the effect be of passing context here? 

i'm going to add a new section to describe the behaviors between with and without webhook(admission controller). how about below description ?


### Behaviors with and without Mutating webhook

Since the mutating webhook is optional for users, we will explain the different behaviors between with and without mutating webhook.

#### with Mutating webhook

**kubectl request**:

- APIServer uses otel to start a new Span
- APIServer uses otel to propagate SpanContext to the other end(webhook)
- Webhook generates an initialtraceid
- Webhook persists SpanContext and initialtraceid to object

**controllers request:**

- Controller uses otel to start a related Span, which connected to the SpanContext in object
- Controller uses otel to propagate Baggage(initialtraceid)  stored in object and SpanContext  to the other end(APIServer)
- APIServer uses otel  to start a related Span, which connected to the SpancContext from the  incoming request
- APIServer uses otel to propagate SpanContext and Baggage(initialtraceid) to the other end(webhook)
- Webhook persists SpanContext and initialtraceid to object

#### without Mutating webhook nor without admission controller

**kubectl request:**

- APIServer uses otel to start a new Span
- APIServer uses otel to propagate SpanContext to the other end
- ~~Webhook generates an initialtraceid~~
- ~~Webhook persists SpanContext and initialtraceid to object~~

**controllers request:**

- Controller start uses otel to start a new Span
- Controller uses otel to propagate SpanContext  to the other end(APIServer)
- APIServer uses otel  to start related Span, which connected to the SpancContext from the  incoming request
- APIServer uses otel to propagate SpanContext to the other end
- ~~Webhook persists SpanContext and initialtraceid to object~~

In short, webhook decides whether logging metadata will be added into object or not.